datafusion/physical-expr/src/aggregate/average.rs

```diff
@@ -383,6 +417,234 @@ impl RowAccumulator for AvgRowAccumulator {
+                        )
+                    });
+
+                let remainder_bits = bit_chunks.remainder_bits();
+                group_indices_remainder
+                    .iter()
+                    .zip(data_remainder.iter())
+                    .enumerate()
+                    .for_each(|(i, (group_index, new_value))| {
+                        if remainder_bits & (1 << i) != 0 {
+                            let sum = &mut self.sums[*group_index];
+                            *sum = sum.add_wrapping(*new_value);
+                        }
+                    });
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<T, F> GroupsAccumulator for AvgGroupsAccumulator<T, F>
+where
+    T: ArrowNumericType + Send,
+    F: Fn(T::Native, u64) -> Result<T::Native> + Send,
+{
+    fn update_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indicies: &[usize],
+        opt_filter: Option<&arrow_array::BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "single argument to update_batch");
+        let values = values.get(0).unwrap().as_primitive::<T>();
+
+        // update counts (TOD account for opt_filter)
+        self.counts.resize(total_num_groups, 0);
+        group_indicies.iter().for_each(|&group_idx| {
+            self.counts[group_idx] += 1;
+        });
+
+        // update values
+        self.update_sums(values, group_indicies, opt_filter, total_num_groups)?;
+
+        Ok(())
+    }
+
+    fn merge_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indicies: &[usize],
+        opt_filter: Option<&arrow_array::BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 2, "two arguments to merge_batch");
+        // first batch is counts, second is partial sums
+        let counts = values.get(0).unwrap().as_primitive::<UInt64Type>();
+        let partial_sums = values.get(1).unwrap().as_primitive::<T>();
+
+        // update counts by summing the partial sums (TODO account for opt_filter)
+        self.counts.resize(total_num_groups, 0);
+        group_indicies.iter().zip(counts.values().iter()).for_each(
+            |(&group_idx, &count)| {
+                self.counts[group_idx] += count;
+            },
+        );
+
+        // update values
+        self.update_sums(partial_sums, group_indicies, opt_filter, total_num_groups)?;
+
+        Ok(())
+    }
+
+    fn evaluate(&mut self) -> Result<ArrayRef> {
+        let counts = std::mem::take(&mut self.counts);
+        let sums = std::mem::take(&mut self.sums);
+
+        let averages: Vec<T::Native> = sums
+            .into_iter()
+            .zip(counts.into_iter())
+            .map(|(sum, count)| (self.avg_fn)(sum, count))
+            .collect::<Result<Vec<_>>>()?;
+
+        // TODO figure out how to do this without the iter / copy
+        let array = PrimitiveArray::<T>::from_iter_values(averages);
```

Comment:
```suggestion
        let array = PrimitiveArray::<T>::from(averages);
```
Should avoid a copy

Reply:
When I try that I get a compiler error:

```rust
   Compiling datafusion-physical-expr v27.0.0 (/Users/alamb/Software/arrow-datafusion/datafusion/physical-expr)
error[E0277]: the trait bound `arrow_array::PrimitiveArray<T>: From<Vec<<T as arrow_array::ArrowPrimitiveType>::Native>>` is not satisfied
   --> datafusion/physical-expr/src/aggregate/average.rs:623:47
    |
623 |         let array = PrimitiveArray::<T>::from(averages);
    |                     ------------------------- ^^^^^^^^ the trait `From<Vec<<T as arrow_array::ArrowPrimitiveType>::Native>>` is not implemented for `arrow_array::PrimitiveArray<T>`
    |                     |
    |                     required by a bound introduced by this call
    |
help: consider extending the `where` clause, but there might be an alternative better way to express this requirement
    |
573 |     F: Fn(T::Native, u64) -> Result<T::Native> + Send, arrow_array::PrimitiveArray<T>: From<Vec<<T as arrow_array::ArrowPrimitiveType>::Native>>
    |                                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I tried following @Dandandan 's suggestion in https://github.com/apache/arrow-datafusion/pull/6800#discussion_r1248737610 but I still get the same error